src/renderers/shared/fiber/ReactFiberBeginWork.js

```diff
@@ -450,7 +457,9 @@ module.exports = function<T, P, I, TI, PI, C, CX, PL>(
       'An indeterminate component should never have mounted. This error is ' +
       'likely caused by a bug in React. Please file an issue.'
     );
-    var fn = workInProgress.type;
+    var fn = hasMockingBehavior
+      ? mockComponent(workInProgress, hostContext.getRootHostContainer())
```

Comment:
@acdlite one problem with this option is that then the resulting data structure in `toTree()` would have `type` be set to the mock, rather than the original value